src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,333 @@
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var ifStatement = root.FindToken(textSpan.Start).GetAncestor<IfStatementSyntax>();
+            if (ifStatement == null || ifStatement.Else == null)
+            {
+                return;
+            }
+
+            if (!ifStatement.IfKeyword.Span.IntersectsWith(textSpan.Start))
+            {
+                return;
+            }
+
+            if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    CSharpFeaturesResources.Invert_if_statement,
+                    c => InvertIfAsync(document, ifStatement, c)));
+        }
+
+        private async Task<Document> InvertIfAsync(Document document, IfStatementSyntax ifStatement, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var ifNode = ifStatement;
+
+            // In the case that the else clause is actually an else if clause, place the if
+            // statement to be moved in a new block in order to make sure that the else
+            // statement matches the right if statement after the edit.
+            var newIfNodeStatement = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement ?
+                SyntaxFactory.Block(ifNode.Else.Statement) :
+                ifNode.Else.Statement;
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var invertedIf = ifNode.WithCondition(Negate(ifNode.Condition, semanticModel, cancellationToken))
+                      .WithStatement(newIfNodeStatement)
+                      .WithElse(ifNode.Else.WithStatement(ifNode.Statement))
+                      .WithAdditionalAnnotations(Formatter.Annotation);
+
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var result = root.ReplaceNode(ifNode, invertedIf);
+            return document.WithSyntaxRoot(result);
+        }
+
+        private bool IsComparisonOfZeroAndSomethingNeverLessThanZero(BinaryExpressionSyntax binaryExpression, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            var canSimplify = false;
+
+            if (binaryExpression.Kind() == SyntaxKind.GreaterThanExpression &&
+                binaryExpression.Right.Kind() == SyntaxKind.NumericLiteralExpression)
+            {
+                canSimplify = CanSimplifyToLengthEqualsZeroExpression(
```

Comment:
several of these branches seem identical (or just swap left/right), can we make this simpler?